### DISCUSSION TITLE
How to avoid having to take devDependencies in production ?

### DISCUSSION THREAD

**Discussion by Yimiprod** [vercel/next.js#36119] [Help]
Hi.

We run a project with Next 10.0.4 and we stumbled upon what we think is an incorrect behavior we don't know how to avoid.

**next.config.js**
```js 
const withPlugins = require('next-compose-plugins');
const withBundleAnalyzer = require('@next/bundle-analyzer');
const withSourceMaps = require('@zeit/next-source-maps');
const withCSS = require('@zeit/next-css');
const LodashModuleReplacementPlugin = require('lodash-webpack-plugin');

const nextConfiguration = {
  assetPrefix: process.env.NEXTJS_ASSET_PREFIX || '',
  onDemandEntries: {
    // period (in ms) where the server will keep pages in the buffer
    maxInactiveAge: 25 * 10000,
    // number of pages that should be kept simultaneously without being disposed
    pagesBufferLength: 10,
  },
  reactStrictMode: true,
  webpack: (config, { dev, isServer, webpack }) => {
    config.plugins.push(new LodashModuleReplacementPlugin());

    return config;
  },
};

module.exports = withPlugins(
  [
    [
      withCSS,
      {
        postcssLoaderOptions: {
          sourceMap: process.env.NODE_ENV !== 'production',
        },
      },
    ],
    [withBundleAnalyzer({ enabled: process.env.ANALYZE === 'true' })],
    [withSourceMaps()],
  ],
  nextConfiguration,
);
```

If we run it with `lodash-webpack-plugin` as a devDependencies after using `yarn install --production` the package is not present and thus making next crash on start.
We had to add it as a dependencies even if it's never used since its in the `webpack()` function which is not called on run.

Is there a more efficient way to define that webpack function to not have to include modules used for building the app in production ?

Edit: same goes for all the others modules by the way, we don't need to use bundleAnalyser, withSourceMaps or withCSS to run the application

**Comment by GabenGar** [comment]
There is literally the `dev` argument in the function you pass. You can [take a gander](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config) on what it could possibly mean.

**Reply by Yimiprod** [reply]
So you mean that i have to put my require inside the `webpack()` function behind the condition ?

**Reply by GabenGar** [reply]
Yes, but probably more than that, considering you have other plugins which don't get invoked inside `webpack()`.
Although I've looked at them and both `@zeit/next-css`  `@zeit/next-source-maps` say they are deprecated, so you probably can drop them from dependencies. 
`next-compose-plugins` looks like a needless wrapper for your usecase (unless it's an example, you don't have enough plugins to warrant it).
`@next/bundle-analyzer` says to import it conditionally which probably matters a lot for CJS:
>Note: if installing as a devDependency make sure to wrap the require in a `process.env` check as next.config.js is loaded during next start as well.

So if you do a deps maintenance work you can drop 4 out of 5 imports in this config and probably move lodash import behind a dev condition too.
